lib/rspec/core/did_you_mean/proximity.rb

```diff
@@ -0,0 +1,102 @@
+module RSpec
+  module Core
+    module DidYouMean
+      # class to calculate jaro_winkler distance based on
```

Comment:
I'm unsure if this is worth this complexity, is there not a similar simpler algorithm we can use?

Reply:
Levenshtein is simpler to implement, might have to benchmark the two for comparison and see what suits better, and whats faster https://github.com/ioquatix/build-text/blob/master/lib/build/text/merge.rb#L60-L102

Reply:
And it's there already in RubyGems:
```
> require 'rubygems/text'
> include Gem::Text
> levenshtein_distance 'abcd', 'sdff' # => 4
```

Reply:
Mm, we can't use that directly, as we don't depend on rubygems and its not always available. We could however make this contextual on that, (e.g. disable the feature on load error.
```
module RSpec
  module Core
    module DidYouMean
      begin
        require 'rubygems/text'
        # ... define did you mean logic
      rescue LoadError
        # ... define identical public entry point which returns the old behaviour
      end
```

Reply:
I am still analysing levenshtein vs jaro-winkler so will comment on the idea of the using the text gem when I have completed the analysis.

Reply:
I have implemented the levenshtein and did a simple benchmark.   When run on the rspec-core files, levenshtein takes about 100ms, whereas jaro-winkler took 17ms.  Levenshtein is clearly O(n^2) where ```n``` is the string length, but according to this [article](http://users.cecs.anu.edu.au/~Peter.Christen/publications/tr-cs-06-02.pdf) jaro-winkler is O(n).   This means if some projects are deeply nested i.e. have longer strings for the file names, then jaro-winkler will have an ever greater speed improvement over levenshtein.  

Just based on my very limited testing, I think jaro-winkler may give better suggestions than levensthein though this may becasause of the cut_offs I have set.

As well, although the levensthein algorithm is shorter, it is a very clever bit of logic that uses dynamic programming to find the minimum number of inserts/deletions/substitutions to transform one string to another. By contrast jaro-winkler is simply counting the number of matches and transpositions so is conceptually much simpler.  As well I think I can simplify the jaro-winkler algorithm by removing the winkler part which gives priority over the first four characters which in our is alway 'spec'.

So I suggest that we go with a simplified jaro-winkler. But I am happy to go with levenshtein, and if we do that I will consider the point about using the text gem.

Reply:
The total time for the did_you_mean is directly proportional to the number of files in the spec folder.  If we go with jaro-winkler this is about 0.2 ms per file.   For 1,000 files this is about 0.2 seconds which would not be perceptible, but for 10,000 files it would be 2 seconds which would cause a marked delay. If you think some users will have more than 1,000 files, I suggest I develop a test to determine the expected time and not look for suggestions if the time is greater than about 500ms.  Please let me know if you think I should do this.

Reply:
As suggested by palkan at #2590 I have changed to using the DidYouMean api.  Using the DidYouMean api, a typical check takes about 6ms on my machine. If I put in a really bad long and unlikely string, it takes  17ms.   On average it is much, much faster than Levenstein and faster than my implementation of jaro-winkler.  It seems to give good suggestions and is much simpler.